src/tools/idlc/src/descriptor.c

```diff
@@ -1287,8 +1289,12 @@ emit_array(
        we're processing a struct type, and this can be used to determine if its
        an external member */
     idl_node_t *parent = idl_parent(node);
-    if (idl_is_struct(stype->node) && idl_is_external(parent))
-      opcode |= DDS_OP_FLAG_EXT;
+    if (idl_is_struct(stype->node)) {
+      if (idl_is_external(parent))
+        opcode |= DDS_OP_FLAG_EXT;
+      if (idl_is_optional(parent))
+        opcode |= DDS_OP_FLAG_OPT;
```

Comment:
See some earlier comment, I am thinking life would be simpler if this were `opcode |= DDS_OP_FLAG_OPT | (is_string ? 0 : DDS_OP_FLAG_EXT)` if the type is not a string. (It should also be documented in the header file that things work this way.)